scipy/spatial/src/views.h

```diff
@@ -0,0 +1,26 @@
+#pragma once
+
+#include <vector>
+#include <array>
+#include <cstdint>
+
+struct ArrayDescriptor {
+    ArrayDescriptor(int64_t ndim):
+        ndim(ndim), shape(ndim, 1), strides(ndim, 0) {
+    }
+
+    int64_t ndim;
+    int64_t element_size;
+    std::vector<int64_t> shape, strides;
+};
+
+template <typename T>
+struct StridedView2D {
```

Comment:
Does py::array support a similar indexing syntax, and what's the motivation for not using it? If it is performance gain of unrolling and vectorizing length-2 loops, then should this optimization be done in pybind11 side rather than on with a workaround on the scipy side?

Reply:
Reasons to use this:
- Doesn't require GIL
- Can do broadcasting/slicing without creating new python objects.

Reply:
Thanks for the explanation. So these benefits are similar to cython's memoryview objects.

Is the unchecked() member of any relevance here?

https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html#direct-access

Reply:
> Is the unchecked() member of any relevance here?

That seems to be for indexing into an existing array. Not for manipulating strides and creating new views on raw data like I'm doing here. Also note that currently the distance functions are all decoupled from pybind11 which is why they could be re-used in ckdtree without issue.

Reply:
My concern (but not against) is that we grow a full different implementation of stride-array interface in C++ inside scipy. The current implementation is simple, but quickly we may start to demand new features when we reuse it. If we take this approach then we'll just keep reusing what you are building here, and make it better along the way.

On the other hand, there might be an underlying mechanism in pybind11 that can be reused for creating new views on raw data. At least this can be done via delegating memory management to a py::array. Also, there isn't much stopping us from rewriting ckdtree to use py::array::unchecked if it can work here.

Reply:
I'm wondering if it would then be worth adopting [xtensor](https://xtensor.readthedocs.io/en/latest/index.html).

Reply:
Replacing 10 lines of code with a new library dependency seems pre-mature. IMO we can look into more advanced strided view types when/if they're needed.

As for pybind11 unchecked references, they're explicitly not user-constructible and instead you need a a `py::array` (NumPy array object) with the shape and strides you want. That's far more heavy-weight than this. Note that both `pdist` and `cdist` directly modify the sizes/strides  multiple times in a loop, so this is performance sensitive.